### ISSUE TITLE
X-axis in LayeredChart does not align correctly upon sorting by another field with Alt.EncodingSortField

### ISSUE THREAD

**Issue by tonylee3399** [altair-viz/altair#2102]
Hi, i'm having a difficulty to correctly sort the order of the X-axis which shows the `Checkpoint` field of the dataset while it is sorted by its `StartTime` field.

```
from io import StringIO
import pandas as pd
import altair as alt

df = pd.read_csv(StringIO("""StartTime	Unit	Value	Checkpoint
0	2020-03-27 03:39:00	A	0.6844	T0
2	2020-03-27 13:57:00	A	0.7092	Check In
3	2020-03-28 18:29:00	A	0.6965	#4
4	2020-03-30 10:34:00	A	0.6896	#7
5	2020-03-30 14:57:00	A	0.7089	#12
6	2020-03-30 16:38:00	A	0.6977	#16
7	2020-03-31 08:01:00	A	0.7076	#18
1	2020-03-27 03:46:00	B	0.6495	T0
8	2020-03-27 13:56:00	B	0.6653	Check In
9	2020-03-27 17:42:00	B	0.6534	#4
10	2020-03-28 16:01:00	B	0.6879	#8
11	2020-03-30 14:47:00	B	0.6903	#12"""), sep='\t')
df.StartTime = pd.to_datetime(df.StartTime, format='%Y-%m-%d %H:%M:%S')
df = df.sort_values(by=['Unit', 'StartTime'])

line = alt.Chart(df).mark_line().encode(
    x = alt.X('Checkpoint:N', sort=alt.EncodingSortField(field='StartTime:T')),
    y = alt.Y('Value:Q', scale=alt.Scale(domain=[0.5, 0.8])),
    color = 'Unit:N',
    tooltip = [alt.Tooltip('StartTime:T', format="%Y-%m-%d %H:%M:%S"), 'Unit', 'Value', 'Checkpoint']
)

proper_line = alt.Chart(df.query('Unit == "A"')).mark_line().encode(
    x = alt.X('Checkpoint:N', sort=alt.EncodingSortField(field='StartTime:T')),
    y = alt.Y('Value:Q', scale=alt.Scale(domain=[0.5, 0.8])),
    color = 'Unit:N',
    tooltip = [alt.Tooltip('StartTime:T', format="%Y-%m-%d %H:%M:%S"), 'Unit', 'Value', 'Checkpoint']
)

proper_line2 = alt.Chart(df.query('Unit == "B"')).mark_line().encode(
    x = alt.X('Checkpoint:N', sort=alt.EncodingSortField(field='StartTime:T')),
    y = alt.Y('Value:Q', scale=alt.Scale(domain=[0.5, 0.8])),
    color = 'Unit:N',
    tooltip = [alt.Tooltip('StartTime:T', format="%Y-%m-%d %H:%M:%S"), 'Unit', 'Value', 'Checkpoint']
)

alt.hconcat(
    (line + line.mark_point()).properties(title='X-axis out of aligned StartTime order'), 
    (proper_line + proper_line.mark_point()).properties(title='X-axis ordered correctly based on StartTime'),
    (proper_line2 + proper_line2.mark_point()).properties(title='X-axis ordered correctly based on StartTime'),
).resolve_legend('independent')
```

The X-axis is ordered correctly if I plot it individually as is shown on the middle and right chart below, and, ordered incorrectly in the `Layered` chart on the left:

![image](https://user-images.githubusercontent.com/16498167/79720153-90c98f80-8312-11ea-99c0-18eed898d9f4.png)

I tried to use the `resolve_axis` method but it seems that it still does not show how I want the data to be plotted chronologically by X-axis. Also, it does not work in my case because the `Unit` variable contains more than 5 unique values, such as: ['A', 'B', 'C', 'D', 'E', ....]

`(line + line.mark_point()).resolve_axis(x='independent')`

Code above resulted in:

![image](https://user-images.githubusercontent.com/16498167/79720303-db4b0c00-8312-11ea-8bdb-0ed50fd2eb86.png)

What I want to achieve is that the `Checkpoint` is ordered based on overall chronological (`StartTime`) order across different `Unit`. In the example above, the `#8` and `#12` Checkpoints are incorrectly placed in the Chart. But if plot them individually, it works perfectly. Any suggestion on this? Thanks!

**Comment by jakevdp** [comment]
You cannot use type shorthands within sortfield. Use this instead:
```
sort=alt.EncodingSortField('StartTime')
```

**Comment by tonylee3399** [comment]
Thanks Jake! I'm glad it's just an error from my end. I was referring to [Sorting Legends](https://altair-viz.github.io/user_guide/encoding.html#sorting) and [Encoding Channel Options](https://altair-viz.github.io/user_guide/encoding.html#encoding-channel-options) docs. I think I might have missed that `shorthand` cannot be used within `EncodingSortField`. I assumed that it could take a shorthand because `sort` options could take `-x` or `-y` for sorting descending.

**Comment by tonylee3399** [comment]
Hi @jakevdp , what is the earliest version that support this? I'm using version 3.2.0 at this moment and it seemed that the X-axis doesn't get ordered by other field correctly.

**Comment by jakevdp** [comment]
There were several bugs related to sorting that were fixed in Altair version 4.0. Please update to a newer version for the fix.
